daprdocs/content/en/java-sdk-docs/spring-boot/_index.md

```diff
@@ -0,0 +1,285 @@
+
+```properties
+dapr.pubsub.name=pubsub
+```
+
+Because with Dapr you can connect to multiple PubSub brokers you need to provide the following bean to let Dapr know which PubSub broker your `DaprMessagingTemplate` will use: 
+```java
+@Bean
+public DaprMessagingTemplate<Order> messagingTemplate(DaprClient daprClient,
+                                                             DaprPubSubProperties daprPubSubProperties) {
+  return new DaprMessagingTemplate<>(daprClient, daprPubSubProperties.getName());
+}
+```
+
+Finally, because Dapr PubSub requires a bidirectional connection between your application and Dapr you need to expand your Testcontainers configuration with a few parameters: 
+
+```java
+@Bean
+@ServiceConnection
+public DaprContainer daprContainer(Network daprNetwork, PostgreSQLContainer<?> postgreSQLContainer, RabbitMQContainer rabbitMQContainer){
+    
+    return new DaprContainer("daprio/daprd:1.14.1")
+            .withAppName("producer-app")
+            .withNetwork(daprNetwork)
+            .withComponent(new Component("kvstore", "state.postgresql", "v1", STATE_STORE_PROPERTIES))
+            .withComponent(new Component("kvbinding", "bindings.postgresql", "v1", BINDING_PROPERTIES))
+            .withComponent(new Component("pubsub", "pubsub.rabbitmq", "v1", rabbitMqProperties))
+            .withAppPort(8080)
+            .withAppChannelAddress("host.testcontainers.internal")
+            .dependsOn(rabbitMQContainer)
+            .dependsOn(postgreSQLContainer);
+}
+```
+
+Now, in the Dapr configuration we have included a `pubsub` component that will connect to an instance of RabbitMQ started by Testcontainers. 
+We have also set two important parameters `.withAppPort(8080)` and `.withAppChannelAddress("host.testcontainers.internal")` which allows Dapr to 
+contact back to the application when a message is published in the broker.
+
+To listen to events/messages you need to expose an endpoint in the application that will be responsible to receive the messages. 
+If you expose a REST endpoint you can use the `@Topic` annotation to let Dapr know where it needs to forward the events/messages too: 
+
+```java
+@PostMapping("subscribe")
+@Topic(pubsubName = "pubsub", name = "topic")
+public void subscribe(@RequestBody CloudEvent<Order> cloudEvent){
+    events.add(cloudEvent);
+}
+```
+
+On bootstrapping your application, Dapr will register the subscription to messages to be forwarded to the `subscribe` endpoint exposed by your application. 
+
+If you are writing tests got these subscribers you need to make sure that let Testcontainers know that your application will be running in port 8080, 
+so containers started with Testcontainers know where your application is: 
+
+```java
+@BeforeAll
+public static void setup(){
+  org.testcontainers.Testcontainers.exposeHostPorts(8080);
+}
+```
+
+You can check and run the [full example source code here](https://github.com/salaboy/dapr-spring-boot-docs-examples).
```

Comment:
We typically link to examples in the java-sdk so that they are maintained and kept updated. Is this example in the sdk and if not, can you add it?

Reply:
This is a bit different, as it uses Testcontainers and the intention here is to keep it as simple as possible for spring users to copy. Maybe we can have a separate directory for it inside the examples, but I wanted to make sure that this doesn't get all the dependencies from other examples that overly complicate it when users want to copy the project structure.